### PR TITLE
Fix segfault when column header is skipped

### PR DESCRIPTION
**Pull request by mdboom**
Hopefully addresses #3545.

### CODE REVIEW

**Comment by mdboom** [comment]
This will probably have a merge conflict with #3525 -- just a note in case we want to merge in a particular order.

**Comment by embray** [comment]
Changelog entry nag, and this can be merged :D

**Comment by mdboom** [comment]
Sure -- I wouldn't mind getting confirmation from @martynbristow that this works for him first, as well.

**Comment by martynbristow** [comment]
Hello

I’m just trying to check it through
I’ve just received a Python(2483,0x7fff7c18c300) malloc: *** error for object 0x100306d10: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6
I’ll see if its something I have done, as I have my real satrapy here too
Martyn Bristow
Mobile Tel: (+44) 7825 165-420
About Me: http://about.me/marty.mcfly <http://about.me/marty.mcfly>
Webpage:- http://www.martynbristow.co.uk <http://www.martynbristow.co.uk/>
E-mail:- martyn.bristow@gmail.com <mailto:martyn.bristow@gmail.com>
E-mail:- martyn@martynbristow.co.uk <mailto:martyn@martynbristow.co.uk>

**Comment by taldcroft** [comment]
@mdboom - works for me, thanks!  After doing this I get the expected exception (because the data file is really malformed for the `tab` format):
```
CParserError: an error occurred while parsing table data: too many columns found in line 1 of data
```

**Comment by martynbristow** [comment]
Yeah brilliant
All fixed, plus it now compensates for stupid text headers ;)

**Comment by embray** [comment]
As feared, this needs to be rebased now.

**Comment by embray** [comment]
@mdboom Can you rebase this?  If you're not around no problem though, I'll just merge by hand.

**Comment by mdboom** [comment]
Rebased.

**Comment by embray** [comment]
Oh, I think this broke as a result of the changes I made to the tests in #3525.

File: astropy/io/ascii/tests/test_c_reader.py [hunk 1]
```diff
@@ -873,3 +873,16 @@ def test_empty_quotes(parallel, read_basic):
     table = read_basic('a b\n1 ""\n2 ""', parallel=parallel)
     expected = Table([[1, 2], [0, 0]], names=('a', 'b'))
     assert_table_equal(table, expected)
+
+@pytest.mark.parametrize("parallel", [True, False])
+def test_fast_tab_with_names(parallel):
```
Comment (embray) [inline astropy/io/ascii/tests/test_c_reader.py]: This needs to be changed to pass in `read_tab` as an argument, as it is now a fixture--somewhat odd I admit.  Sorry 'bout that.

**Comment by embray** [comment]
FWIW I fixed that up and rebased it--the new branch is testing here: https://travis-ci.org/embray/astropy/builds/53374609

Looks like it should pass, then I'll merge.

**Comment by embray** [comment]
Manually merged in 00d223f.

**Comment by mdboom** [comment]
Thanks!

**Comment by taldcroft** [comment]
Thanks both!!

**Pull request closed without merging**